osd/PGPool::update: optimize with subset_of

zmedico opened the pull request:
Replace expensive inverval_set intersection_of and operator==
calls with a single subset_of call. I borrowed this idea from
Piotr Dałek's "osd/PGPool: don't use intermediate interval set"
patch. The following benchmark program demonstrates a 38%
performance increase:
```c++
#include <iostream>
#include <chrono>
#include "include/interval_set.h"

#define NANOSECONDS(d) \
    std::chrono::duration_cast<std::chrono::nanoseconds>(d).count()

typedef uint64_t snapid_t;
typedef std::chrono::steady_clock::duration duration;

duration PGPool_update_old(const interval_set<snapid_t> &rs) {
  std::chrono::steady_clock::time_point start, end;
  interval_set<snapid_t> newly_removed_snaps, cached_removed_snaps;

  // initialize state
  cached_removed_snaps = rs;

  // start timed simulation
  start = std::chrono::steady_clock::now();

  {
    newly_removed_snaps = cached_removed_snaps;
    interval_set<snapid_t> intersection;
    intersection.intersection_of(newly_removed_snaps, cached_removed_snaps);

    assert(intersection == cached_removed_snaps);
    cached_removed_snaps.swap(newly_removed_snaps);
    newly_removed_snaps = cached_removed_snaps;
    newly_removed_snaps.subtract(intersection);
  }

  // end timed simulation
  end = std::chrono::steady_clock::now();

  return end - start;
}

duration PGPool_update_new(const interval_set<snapid_t> &rs) {
  std::chrono::steady_clock::time_point start, end;
  interval_set<snapid_t> newly_removed_snaps, cached_removed_snaps;

  // initialize state
  cached_removed_snaps = rs;

  // start timed simulation
  start = std::chrono::steady_clock::now();

  {
    newly_removed_snaps = cached_removed_snaps;

    assert(cached_removed_snaps.subset_of(newly_removed_snaps));
    interval_set<snapid_t> removed_snaps = newly_removed_snaps;
    newly_removed_snaps.subtract(cached_removed_snaps);
    cached_removed_snaps.swap(removed_snaps);
  }

  // end timed simulation
  end = std::chrono::steady_clock::now();

  return end - start;
}

int main(int argc, char *argv[])
{
  assert(argc == 3);
  const int sample_count = std::stoi(argv[1]);
  const int interval_count = std::stoi(argv[2]);
  const int interval_distance = 4;
  const int interval_size = 2;
  const int max_offset = interval_count * interval_distance;
  interval_set<snapid_t> removed_snaps;

  for (int i = 0; i < max_offset; i += interval_distance)
    removed_snaps.insert(i, interval_size);

  duration old_delta(0), new_delta(0);

  for (int i = 0; i < sample_count; ++i) {
    old_delta += PGPool_update_old(removed_snaps);
    new_delta += PGPool_update_new(removed_snaps);
  }

  float ratio = float(NANOSECONDS(old_delta)) / NANOSECONDS(new_delta);

  std::cout << ratio << std::endl;
}
```
@branch-predictor

zmedico commented:
I need to benchmark this. It looks like subset_of might need to be optimized, since it calls contains, which uses the find_inc/lower_bound method.

zmedico commented:
I've added an optimized subset_of implementation, and my benchmark program is showing a 38% improvement in performance.

zmedico commented:
I've updated unittest_interval_set to cover all cases of the `subset_size_sym` function.

gregsfortytwo commented on src/include/interval_set.h:
```diff
@@ -286,6 +286,41 @@ class interval_set {
+    while (pa != a_end && pb != b_end) {
+
+      if (pb->first + pb->second <= pa->first)
+        { ++pb;  continue; }
```
This will break if you give it input where pb runs through to the end.

zmedico replied:
The `pb != b_end` bound is checked by the outer while loop. I did think about using an inner while loop here, but it didn't seem worth it since in only optimizes away one `pa != a_end` comparison.

zmedico replied:
I'll go ahead and change it to a while loop with bounds check.

zmedico replied:
Updated as follows:
```c++
      while (pb->first + pb->second <= pa->first) {
        ++pb;
        if (pb == b_end)
          return false;
      }
```

zmedico commented:
Added the following bounds check:
```c++
      while (pb->first + pb->second <= pa->first) {
        ++pb;
        if (pb == b_end)
          return false;
      }
```

gregsfortytwo commented:
Let's get it into the next testing run somebody does as-is though, in case I'm missing something.

zmedico commented on src/include/interval_set.h:
```diff
@@ -286,6 +286,44 @@ class interval_set {
+      }
+
+      if (pa->first < pb->first)
+        return false;
```
Both `pa->first < pb->first` and `pa->first + pa->second <= pb->first`  are distinct cases that need to be handled.

`pa->first < pb->first` means the shorter set's interval begins before the other interval

`pa->first + pa->second <= pb->first` means the shorter set's interval ends before the other interval begins

So, there's no redundancy in the existing logic. It's tricky enough to visualize that maybe we should have comments for each case, like:
```c++
      // interval begins before other
      if (pa->first < pb->first)
        return false;

      // interval ends before other begins
      if (pa->first + pa->second <= pb->first)
        return false;

      // interval is longer than other
      if (pa->first + pa->second > pb->first + pb->second)
        return false;
```
I'd also like to add an earlier range_end() comparison back in the subset_of method.

zmedico replied:
I see it now, "interval ends before other begins" is a case of "interval begins before other", so I've removed the redundant case.

zmedico commented on src/include/interval_set.h:
```diff
@@ -286,6 +286,44 @@ class interval_set {
+                                               b_end = b.m.end();
+
+    while (pa != a_end && pb != b_end) {
+
```
Okay, I've removed some whitespace in the latest update. Let me know how it looks.

zmedico commented:
Updates:
* Removed some whitespace
* Re-ordered logic in subset_size_sym (no change in behavior), and added comments
* Added early `big.range_end() < range_end()` check to subset_of

tchaikov commented on src/include/interval_set.h:
```diff
@@ -286,6 +286,43 @@ class interval_set {
   }
+
+  bool subset_size_sym(const interval_set &b) const {
+    typename decltype(m)::const_iterator pa = m.begin(),
```
might want to use
```c++
auto pa = m.cbegin(), pb = b.m.cbegin();
```

zmedico replied:
Thanks, done.

gregsfortytwo commented:
Okay, I believe the docs issue was in master and fixed in master, and this looks great. Just pending a testing branch run with it.

zmedico commented:
Rebased on changes from caf6803b13d9dbd4540da366b018d721fcfc371a.

yuriw commented:
RBD
http://pulpito.ceph.com/yuriw-2017-09-27_15:26:37-rbd-wip-yuri-testing4-2017-09-27-1402-distro-basic-smithi/

http://pulpito.ceph.com/yuriw-2017-09-27_20:29:25-rbd-wip-yuri-testing4-2017-09-27-1402-distro-basic-smithi/

RADOS

http://pulpito.ceph.com/yuriw-2017-09-27_15:37:08-rados-wip-yuri-testing4-2017-09-27-1402-distro-basic-smithi/

http://pulpito.ceph.com/yuriw-2017-09-28_02:30:38-rados-wip-yuri-testing4-2017-09-27-1402-distro-basic-smithi/
